platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/codecs/BindingsBackedCodec.kt

```diff
@@ -62,22 +62,42 @@ class BindingsBackedCodec(private val bindings: List<Binding>) : Codec<Any?> {
             }, value) {
                 encoding.run { encode(value) }
             }
+            if (isIntegrityCheckEnabled) {
+                writeSmallInt(tag)
```

Comment:
💅 

Maybe write a tag-derived value instead of the tag value (10000 + tag, for instance) so it we call tell start and end tags apart.

Reply:
I'd be against mixing magic constants and computation.

Since this logic is already behind an opt-in flag, we can write a magic constant as a separate bit, right before the closing tag. It would also be nice to make this constant very unlikely or impossible to match a tag, so our integrity guarantees would be higher.

@mlopatkin, are there scenarios that are not unlikely and in which the corrupted value matches the tag? I imagine it could be possible if the tags are small integers. If such scenarios exist, we would read the something else that we would take for a closing tag and misinform the user about what is actually corrupted.

Reply:
@mlopatkin Does it mean you prefer keeping it as it is now -- writing only the same-value-closing-tag?
It _is_ about the probability, and I wonder if it's low/high enough, when we write the tag plainly